packages/hash/frontend/src/lib/subgraph.ts

```diff
@@ -386,3 +386,29 @@ export const getEntityTypesByBaseUri = (
     return vertex.inner;
   });
 };
+
+// ------------------- Checked Subgraph Objects -------------------
+
+export type SingleEntityRootedSubgraph = Subgraph & { root: Entity };
+
+/**
+ * Checks if the `subgraph` is rooted at a single `Entity` and modifies the object to store a reference to it as
+ * a `root` property
+ * @param subgraph
+ * @throws if there were more or less than one root
+ * @throws if the root wasn't an `EntityVertex`
+ */
+export const isSingleEntityRootedSubgraph = (
+  subgraph: Subgraph,
+): subgraph is SingleEntityRootedSubgraph => {
+  if (subgraph.roots.length !== 1) {
+    throw new Error(
+      `expected subgraph to have a single root but had ${subgraph.roots.length}`,
+    );
+  }
+
+  // eslint-disable-next-line no-param-reassign
+  (subgraph as SingleEntityRootedSubgraph).root = rootsAsEntities(subgraph)[0]!;
```

Comment:
I appreciate assigning here may be a bad idea but I wasn't sure what else to do. 

If we did something like
```ts
intoSingleEntityRootedSubgraph(subgraph: Subgraph): SingleEntityRootedSubgraph
```
then calling it would be something like

```ts
const subgraph: Subgraph = ...;
const newSubgraph = intoSingleEntityRootedSubgraph(subgraph);
```

But I think this is even more misleading because it modifies the original one even though you're getting a new return value. We could get around that by not returning something, but then there's no change in Type and you're somewhat back to where you started.

We could return a new object _without_ modifying the old one but that would be a copy and copying a subgraph could potentially be expensive (although perhaps a shallow copy would be fine?)

cc: @CiaranMn, I think these are the ergonomics we really need to figure out and I'm not sure if this PR is the place to do it or not. Maybe it is?

Reply:
cc @benwerner01

Reply:
I mean that is what I've done, so I'm happy to rename `SingleEntityRootedSubgraph` to `SubgraphAndEntityRoot` if you think that's less confusing.

Reply:
You haven't, you have:
```typescript
type SingleEntityRootedSubgraph = Subgraph & { root: Entity };
```
I am suggesting you have
```typescript
type SubgraphAndEntityRoot = {
  subgraph: Subgraph;
  root: Entity;
}
```
then the process of creating `SubgraphAndEntityRoot` does not involve modifying `Subgraph`

Reply:
> 1. Have a function called getSubgraphRoot that returns the root from a subgraph and leaves it up to the caller what to do with it

I have a function to get the roots of a graph and it leads the caller up to it. 

If you look at the commit, that is used in this function here, and the point of this function is just to store the fact that this check has been made, so we don't have ~10 files constantly looking up the root and checking there's only one and checking that it is an entity. (Hence introduction of `SingleEntityRootedSubgraph`, or if we rename it, `SubgraphAndEntityRoot`)

I appreciate the type-guard is confusing but I'm not sure what the best plan is then because of the concerns with the option I made above

Reply:
> You haven't, you have:

Sorry I did miss that, thanks for the next code snippet, much clearer!

Reply:
Is this what you had in mind @CiaranMn 
https://github.com/hashintel/hash/pull/1242/commits/d91204a4e95303ce6d890f86438267532fe3761c

Reply:
Yes, looks good, thanks!